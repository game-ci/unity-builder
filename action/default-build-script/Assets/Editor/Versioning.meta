fileFormatVersion: 2
guid: 1c3bddf6d8984cde9208e3f0fe584879
timeCreated: 1587490700