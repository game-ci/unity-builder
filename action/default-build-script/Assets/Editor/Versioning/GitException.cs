using System;

namespace UnityBuilderAction.Versioning
{
  public class GitException : InvalidOperationException
  {
    public readonly int code;

    public GitException(int code, string errors) : base(errors)
    {
      this.code = code;
    }
  }
}
