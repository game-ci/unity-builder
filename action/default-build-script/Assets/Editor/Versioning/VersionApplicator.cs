using System;
using JetBrains.Annotations;
using UnityEditor;

namespace UnityBuilderAction.Versioning
{
  public class VersionApplicator
  {
    public static void SetVersion(string version)
    {
      if (version == "none") {
        return;
      }
      
      Apply(version);
    }

    static void Apply(string version)
    {
      PlayerSettings.bundleVersion = version;
    }
  }
}
