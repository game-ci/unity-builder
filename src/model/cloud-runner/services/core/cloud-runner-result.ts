import BuildParameters from '../../../build-parameters';

class CloudRunnerResult {
  public BuildParameters: BuildParameters;
  public BuildResults: string;
  public BuildSucceeded: boolean;
  public BuildFinished: boolean;
  public LibraryCacheUsed: boolean;

  public constructor(
    buildParameters: BuildParameters,
    buildResults: string,
    buildSucceeded: boolean,
    buildFinished: boolean,
    libraryCacheUsed: boolean,
  ) {
    this.BuildParameters = buildParameters;
    this.BuildResults = buildResults;
    this.BuildSucceeded = buildSucceeded;
    this.BuildFinished = buildFinished;
    this.LibraryCacheUsed = libraryCacheUsed;
  }
}
export default CloudRunnerResult;
