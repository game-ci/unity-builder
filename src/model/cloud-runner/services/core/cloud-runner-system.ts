import { exec } from 'child_process';
import { RemoteClientLogger } from '../../remote-client/remote-client-logger';

export class CloudRunnerSystem {
  public static async RunAndReadLines(command: string): Promise<string[]> {
    const result = await CloudRunnerSystem.Run(command, false, true);

    return result
      .split(`\n`)
      .map((x) => x.replace(`\r`, ``))
      .filter((x) => x !== ``)
      .map((x) => {
        const lineValues = x.split(` `);

        return lineValues[lineValues.length - 1];
      });
  }

  public static async Run(command: string, suppressError = false, suppressLogs = false) {
    for (const element of command.split(`\n`)) {
      if (!suppressLogs) {
        RemoteClientLogger.log(element);
      }
    }

    return await new Promise<string>((promise, throwError) => {
      let output = '';
      const child = exec(command, (error, stdout, stderr) => {
        if (!suppressError && error) {
          RemoteClientLogger.log(error.toString());
          throwError(error);
        }
        if (stderr) {
          const diagnosticOutput = `${stderr.toString()}`;
          if (!suppressLogs) {
            RemoteClientLogger.logCliDiagnostic(diagnosticOutput);
          }
          output += diagnosticOutput;
        }
        const outputChunk = `${stdout}`;
        output += outputChunk;
      });
      child.on('close', (code) => {
        if (!suppressLogs) {
          RemoteClientLogger.log(`[${code}]`);
        }
        if (code !== 0 && !suppressError) {
          throwError(output);
        }
        const outputLines = output.split(`\n`);
        for (const element of outputLines) {
          if (!suppressLogs) {
            RemoteClientLogger.log(element);
          }
        }
        promise(output);
      });
    });
  }
}
