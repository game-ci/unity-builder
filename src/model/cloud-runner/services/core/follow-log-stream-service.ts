import GitHub from '../../../github';
import CloudRunner from '../../cloud-runner';
import { CloudRunnerStatics } from '../../options/cloud-runner-statics';
import CloudRunnerLogger from './cloud-runner-logger';
import * as core from '@actions/core';

export class FollowLogStreamService {
  static Reset() {
    FollowLogStreamService.DidReceiveEndOfTransmission = false;
  }
  static errors = ``;
  public static DidReceiveEndOfTransmission = false;
  public static handleIteration(message: string, shouldReadLogs: boolean, shouldCleanup: boolean, output: string) {
    if (message.includes(`---${CloudRunner.buildParameters.logId}`)) {
      CloudRunnerLogger.log('End of log transmission received');
      FollowLogStreamService.DidReceiveEndOfTransmission = true;
      shouldReadLogs = false;
    } else if (message.includes('Rebuilding Library because the asset database could not be found!')) {
      GitHub.updateGitHubCheck(`Library was not found, importing new Library`, ``);
      core.warning('LIBRARY NOT FOUND!');
      core.setOutput('library-found', 'false');
    } else if (message.includes('Build succeeded')) {
      GitHub.updateGitHubCheck(`Build succeeded`, `Build succeeded`);
      core.setOutput('build-result', 'success');
    } else if (message.includes('Build fail')) {
      GitHub.updateGitHubCheck(
        `Build failed\n${FollowLogStreamService.errors}`,
        `Build failed`,
        `failure`,
        `completed`,
      );
      core.setOutput('build-result', 'failed');
      core.setFailed('unity build failed');
      core.error('BUILD FAILED!');
    } else if (message.toLowerCase().includes('error ')) {
      core.error(message);
      FollowLogStreamService.errors += `\n${message}`;
    } else if (message.toLowerCase().includes('error: ')) {
      core.error(message);
      FollowLogStreamService.errors += `\n${message}`;
    } else if (message.toLowerCase().includes('command failed: ')) {
      FollowLogStreamService.errors += `\n${message}`;
    } else if (message.toLowerCase().includes('invalid ')) {
      FollowLogStreamService.errors += `\n${message}`;
    } else if (message.toLowerCase().includes('incompatible  ')) {
      FollowLogStreamService.errors += `\n${message}`;
    } else if (message.toLowerCase().includes('cannot be found')) {
      FollowLogStreamService.errors += `\n${message}`;
    }
    if (CloudRunner.buildParameters.cloudRunnerDebug) {
      output += `${message}\n`;
    }
    CloudRunnerLogger.log(`[${CloudRunnerStatics.logPrefix}] ${message}`);

    return { shouldReadLogs, shouldCleanup, output };
  }
}
