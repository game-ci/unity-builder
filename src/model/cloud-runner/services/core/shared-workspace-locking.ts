import { CloudRunnerSystem } from './cloud-runner-system';
import fs from 'node:fs';
import CloudRunnerLogger from './cloud-runner-logger';
import BuildParameters from '../../../build-parameters';
import CloudRunner from '../../cloud-runner';
export class SharedWorkspaceLocking {
  public static get workspaceBucketRoot() {
    return `s3://${CloudRunner.buildParameters.awsStackName}/`;
  }
  public static get workspaceRoot() {
    return `${SharedWorkspaceLocking.workspaceBucketRoot}locks/`;
  }
  public static async GetAllWorkspaces(buildParametersContext: BuildParameters): Promise<string[]> {
    if (!(await SharedWorkspaceLocking.DoesCacheKeyTopLevelExist(buildParametersContext))) {
      return [];
    }

    return (
      await SharedWorkspaceLocking.ReadLines(
        `aws s3 ls ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/`,
      )
    )
      .map((x) => x.replace(`/`, ``))
      .filter((x) => x.endsWith(`_workspace`))
      .map((x) => x.split(`_`)[1]);
  }
  public static async DoesCacheKeyTopLevelExist(buildParametersContext: BuildParameters) {
    try {
      const rootLines = await SharedWorkspaceLocking.ReadLines(
        `aws s3 ls ${SharedWorkspaceLocking.workspaceBucketRoot}`,
      );
      const lockFolderExists = rootLines.map((x) => x.replace(`/`, ``)).includes(`locks`);

      if (lockFolderExists) {
        const lines = await SharedWorkspaceLocking.ReadLines(`aws s3 ls ${SharedWorkspaceLocking.workspaceRoot}`);

        return lines.map((x) => x.replace(`/`, ``)).includes(buildParametersContext.cacheKey);
      } else {
        return false;
      }
    } catch {
      return false;
    }
  }

  public static NewWorkspaceName() {
    return `${CloudRunner.retainedWorkspacePrefix}-${CloudRunner.buildParameters.buildGuid}`;
  }
  public static async GetAllLocksForWorkspace(
    workspace: string,
    buildParametersContext: BuildParameters,
  ): Promise<string[]> {
    if (!(await SharedWorkspaceLocking.DoesWorkspaceExist(workspace, buildParametersContext))) {
      return [];
    }

    return (
      await SharedWorkspaceLocking.ReadLines(
        `aws s3 ls ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/`,
      )
    )
      .map((x) => x.replace(`/`, ``))
      .filter((x) => x.includes(workspace) && x.endsWith(`_lock`));
  }
  public static async GetLockedWorkspace(workspace: string, runId: string, buildParametersContext: BuildParameters) {
    if (buildParametersContext.maxRetainedWorkspaces === 0) {
      return false;
    }

    if (await SharedWorkspaceLocking.DoesCacheKeyTopLevelExist(buildParametersContext)) {
      const workspaces = await SharedWorkspaceLocking.GetFreeWorkspaces(buildParametersContext);
      CloudRunnerLogger.log(`run agent ${runId} is trying to access a workspace, free: ${JSON.stringify(workspaces)}`);
      for (const element of workspaces) {
        const lockResult = await SharedWorkspaceLocking.LockWorkspace(element, runId, buildParametersContext);
        CloudRunnerLogger.log(
          `run agent: ${runId} try lock workspace: ${element} locking attempt result: ${lockResult}`,
        );

        if (lockResult) {
          return true;
        }
      }
    }

    if (await SharedWorkspaceLocking.DoesWorkspaceExist(workspace, buildParametersContext)) {
      workspace = SharedWorkspaceLocking.NewWorkspaceName();
      CloudRunner.lockedWorkspace = workspace;
    }

    const createResult = await SharedWorkspaceLocking.CreateWorkspace(workspace, buildParametersContext);
    const lockResult = await SharedWorkspaceLocking.LockWorkspace(workspace, runId, buildParametersContext);
    CloudRunnerLogger.log(
      `run agent ${runId} didn't find a free workspace so created: ${workspace} createWorkspaceSuccess: ${createResult} Lock:${lockResult}`,
    );

    return createResult && lockResult;
  }

  public static async DoesWorkspaceExist(workspace: string, buildParametersContext: BuildParameters) {
    return (
      (await SharedWorkspaceLocking.GetAllWorkspaces(buildParametersContext)).filter((x) => x.includes(workspace))
        .length > 0
    );
  }
  public static async HasWorkspaceLock(
    workspace: string,
    runId: string,
    buildParametersContext: BuildParameters,
  ): Promise<boolean> {
    const locks = (await SharedWorkspaceLocking.GetAllLocksForWorkspace(workspace, buildParametersContext))
      .map((x) => {
        return {
          name: x,
          timestamp: Number(x.split(`_`)[0]),
        };
      })
      .sort((x) => x.timestamp);
    const lockMatches = locks.filter((x) => x.name.includes(runId));
    const includesRunLock = lockMatches.length > 0 && locks.indexOf(lockMatches[0]) === 0;
    CloudRunnerLogger.log(
      `Checking has workspace lock, runId: ${runId}, workspace: ${workspace}, success: ${includesRunLock} \n- Num of locks created by Run Agent: ${
        lockMatches.length
      } Num of Locks: ${locks.length}, Time ordered index for Run Agent: ${locks.indexOf(lockMatches[0])} \n \n`,
    );

    return includesRunLock;
  }

  public static async GetFreeWorkspaces(buildParametersContext: BuildParameters): Promise<string[]> {
    const result: string[] = [];
    const workspaces = await SharedWorkspaceLocking.GetAllWorkspaces(buildParametersContext);
    for (const element of workspaces) {
      const isLocked = await SharedWorkspaceLocking.IsWorkspaceLocked(element, buildParametersContext);
      const isBelowMax = await SharedWorkspaceLocking.IsWorkspaceBelowMax(element, buildParametersContext);
      CloudRunnerLogger.log(`workspace ${element} locked:${isLocked} below max:${isBelowMax}`);
      if (!isLocked && isBelowMax) {
        result.push(element);
      }
    }

    return result;
  }

  public static async IsWorkspaceBelowMax(
    workspace: string,
    buildParametersContext: BuildParameters,
  ): Promise<boolean> {
    const workspaces = await SharedWorkspaceLocking.GetAllWorkspaces(buildParametersContext);
    if (workspace === ``) {
      return (
        workspaces.length < buildParametersContext.maxRetainedWorkspaces ||
        buildParametersContext.maxRetainedWorkspaces === 0
      );
    }
    const ordered: any[] = [];
    for (const ws of workspaces) {
      ordered.push({
        name: ws,
        timestamp: await SharedWorkspaceLocking.GetWorkspaceTimestamp(ws, buildParametersContext),
      });
    }
    ordered.sort((x) => x.timestamp);
    const matches = ordered.filter((x) => x.name.includes(workspace));
    const isWorkspaceBelowMax =
      matches.length > 0 &&
      (ordered.indexOf(matches[0]) < buildParametersContext.maxRetainedWorkspaces ||
        buildParametersContext.maxRetainedWorkspaces === 0);

    return isWorkspaceBelowMax;
  }

  public static async GetWorkspaceTimestamp(
    workspace: string,
    buildParametersContext: BuildParameters,
  ): Promise<Number> {
    if (workspace.split(`_`).length > 0) {
      return Number(workspace.split(`_`)[1]);
    }

    if (!(await SharedWorkspaceLocking.DoesWorkspaceExist(workspace, buildParametersContext))) {
      throw new Error("Workspace doesn't exist, can't call get all locks");
    }

    return (
      await SharedWorkspaceLocking.ReadLines(
        `aws s3 ls ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/`,
      )
    )
      .map((x) => x.replace(`/`, ``))
      .filter((x) => x.includes(workspace) && x.endsWith(`_workspace`))
      .map((x) => Number(x))[0];
  }

  public static async IsWorkspaceLocked(workspace: string, buildParametersContext: BuildParameters): Promise<boolean> {
    if (!(await SharedWorkspaceLocking.DoesWorkspaceExist(workspace, buildParametersContext))) {
      throw new Error(`workspace doesn't exist ${workspace}`);
    }
    const files = await SharedWorkspaceLocking.ReadLines(
      `aws s3 ls ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/`,
    );

    const lockFilesExist =
      files.filter((x) => {
        return x.includes(workspace) && x.endsWith(`_lock`);
      }).length > 0;

    return lockFilesExist;
  }

  public static async CreateWorkspace(workspace: string, buildParametersContext: BuildParameters): Promise<boolean> {
    if (await SharedWorkspaceLocking.DoesWorkspaceExist(workspace, buildParametersContext)) {
      throw new Error(`${workspace} already exists`);
    }
    const timestamp = Date.now();
    const file = `${timestamp}_${workspace}_workspace`;
    fs.writeFileSync(file, '');
    await CloudRunnerSystem.Run(
      `aws s3 cp ./${file} ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/${file}`,
      false,
      true,
    );
    fs.rmSync(file);

    const workspaces = await SharedWorkspaceLocking.GetAllWorkspaces(buildParametersContext);

    CloudRunnerLogger.log(`All workspaces ${workspaces}`);
    if (!(await SharedWorkspaceLocking.IsWorkspaceBelowMax(workspace, buildParametersContext))) {
      CloudRunnerLogger.log(`Workspace is above max ${workspaces} ${buildParametersContext.maxRetainedWorkspaces}`);
      await SharedWorkspaceLocking.CleanupWorkspace(workspace, buildParametersContext);

      return false;
    }

    return true;
  }

  public static async LockWorkspace(
    workspace: string,
    runId: string,
    buildParametersContext: BuildParameters,
  ): Promise<boolean> {
    const existingWorkspace = workspace.endsWith(`_workspace`);
    const ending = existingWorkspace ? workspace : `${workspace}_workspace`;
    const file = `${Date.now()}_${runId}_${ending}_lock`;
    fs.writeFileSync(file, '');
    await CloudRunnerSystem.Run(
      `aws s3 cp ./${file} ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/${file}`,
      false,
      true,
    );
    fs.rmSync(file);

    const hasLock = await SharedWorkspaceLocking.HasWorkspaceLock(workspace, runId, buildParametersContext);

    if (hasLock) {
      CloudRunner.lockedWorkspace = workspace;
    } else {
      await CloudRunnerSystem.Run(
        `aws s3 rm ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/${file}`,
        false,
        true,
      );
    }

    return hasLock;
  }

  public static async ReleaseWorkspace(
    workspace: string,
    runId: string,
    buildParametersContext: BuildParameters,
  ): Promise<boolean> {
    const files = await SharedWorkspaceLocking.GetAllLocksForWorkspace(workspace, buildParametersContext);
    const file = files.find((x) => x.includes(workspace) && x.endsWith(`_lock`) && x.includes(runId));
    CloudRunnerLogger.log(`All Locks ${files} ${workspace} ${runId}`);
    CloudRunnerLogger.log(`Deleting lock ${workspace}/${file}`);
    CloudRunnerLogger.log(`rm ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/${file}`);
    await CloudRunnerSystem.Run(
      `aws s3 rm ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey}/${file}`,
      false,
      true,
    );

    return !(await SharedWorkspaceLocking.HasWorkspaceLock(workspace, runId, buildParametersContext));
  }

  public static async CleanupWorkspace(workspace: string, buildParametersContext: BuildParameters) {
    await CloudRunnerSystem.Run(
      `aws s3 rm ${SharedWorkspaceLocking.workspaceRoot}${buildParametersContext.cacheKey} --exclude "*" --include "*_${workspace}_*"`,
      false,
      true,
    );
  }

  public static async ReadLines(command: string): Promise<string[]> {
    return CloudRunnerSystem.RunAndReadLines(command);
  }
}

export default SharedWorkspaceLocking;
