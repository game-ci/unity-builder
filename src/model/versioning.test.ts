import * as core from '@actions/core';
import NotImplementedException from './error/not-implemented-exception';
import System from './system';
import Versioning from './versioning';

afterEach(() => {
  jest.restoreAllMocks();
});

describe('Versioning', () => {
  describe('strategies', () => {
    it('returns an object', () => {
      expect(typeof Versioning.strategies).toStrictEqual('object');
    });

    it('has items', () => {
      expect(Object.values(Versioning.strategies).length).toBeGreaterThan(2);
    });

    it('has an opt out option', () => {
      expect(Versioning.strategies).toHaveProperty('None');
    });

    it('has the semantic option', () => {
      expect(Versioning.strategies).toHaveProperty('Semantic');
    });

    it('has a strategy for tags', () => {
      expect(Versioning.strategies).toHaveProperty('Tag');
    });

    it('has an option that allows custom input', () => {
      expect(Versioning.strategies).toHaveProperty('Custom');
    });
  });

  describe('branch', () => {
    it('returns headRef when set', () => {
      const headReference = jest.spyOn(Versioning, 'headRef', 'get').mockReturnValue('feature-branch-1');

      expect(Versioning.branch).toStrictEqual('feature-branch-1');
      expect(headReference).toHaveBeenCalledTimes(1);
    });

    it('returns part of Ref when set', () => {
      jest.spyOn(Versioning, 'headRef', 'get').mockImplementation();
      const reference = jest.spyOn(Versioning, 'ref', 'get').mockReturnValue('refs/heads/feature-branch-2');

      expect(Versioning.branch).toStrictEqual('feature-branch-2');
      expect(reference).toHaveBeenCalledTimes(2);
    });

    it('prefers headRef over ref when set', () => {
      const headReference = jest.spyOn(Versioning, 'headRef', 'get').mockReturnValue('feature-branch-1');
      const reference = jest.spyOn(Versioning, 'ref', 'get').mockReturnValue('refs/heads/feature-2');

      expect(Versioning.branch).toStrictEqual('feature-branch-1');
      expect(headReference).toHaveBeenCalledTimes(1);
      expect(reference).toHaveBeenCalledTimes(0);
    });

    it('returns undefined when headRef and ref are not set', () => {
      const headReference = jest.spyOn(Versioning, 'headRef', 'get').mockImplementation();
      const reference = jest.spyOn(Versioning, 'ref', 'get').mockImplementation();

      expect(Versioning.branch).not.toBeDefined();

      expect(headReference).toHaveBeenCalledTimes(1);
      expect(reference).toHaveBeenCalledTimes(1);
    });
  });

  describe('headRef', () => {
    it('does not throw', () => {
      expect(() => Versioning.headRef).not.toThrow();
    });
  });

  describe('ref', () => {
    it('does not throw', () => {
      expect(() => Versioning.ref).not.toThrow();
    });
  });

  describe('isDirtyAllowed', () => {
    it('does not throw', () => {
      expect(() => Versioning.isDirtyAllowed).not.toThrow();
    });

    it('returns false by default', () => {
      expect(Versioning.isDirtyAllowed).toStrictEqual(false);
    });
  });

  describe('logging git diff', () => {
    it('calls git diff', async () => {
      // allowDirtyBuild: true
      jest.spyOn(core, 'getInput').mockReturnValue('true');
      jest.spyOn(Versioning, 'isShallow').mockResolvedValue(true);
      jest.spyOn(Versioning, 'isDirty').mockResolvedValue(false);
      jest.spyOn(Versioning, 'fetch').mockImplementation();
      jest.spyOn(Versioning, 'hasAnyVersionTags').mockResolvedValue(true);
      jest
        .spyOn(Versioning, 'parseSemanticVersion')
        .mockResolvedValue({ match: '', tag: 'mocktag', commits: 'abcdef', hash: '75822BCAF' });
      const logDiffSpy = jest.spyOn(Versioning, 'logDiff');
      const gitSpy = jest.spyOn(System, 'run').mockImplementation();

      await Versioning.generateSemanticVersion();

      expect(logDiffSpy).toHaveBeenCalledTimes(1);
      expect(gitSpy).toHaveBeenCalledTimes(1);
      // Todo - this no longer works since typescript
      // const issuedCommand = System.run.mock.calls[0][2].input.toString();
      // expect(issuedCommand.indexOf('diff')).toBeGreaterThan(-1);
    });
  });

  describe('descriptionRegex', () => {
    it('is a valid regex', () => {
      expect(Versioning.descriptionRegex1).toBeInstanceOf(RegExp);
    });

    test.each(['v1.1-1-g12345678', 'v0.1-2-g12345678', 'v0.0-500-gA9B6C3D0-dirty'])(
      'is happy with valid %s',
      (description) => {
        expect(Versioning.descriptionRegex1.test(description)).toBeTruthy();
      },
    );

    test.each(['v0', 'v0.1', 'v0.1.2', 'v0.1-2', 'v0.1-2-g'])('does not like %s', (description) => {
      expect(Versioning.descriptionRegex1.test(description)).toBeFalsy();
      // Also never expect without the v to work for any of these cases.
      expect(Versioning.descriptionRegex1.test(description?.slice(1))).toBeFalsy();
    });
  });

  describe('determineVersion', () => {
    test.each(['somethingRandom'])('throws for invalid strategy %s', async (strategy) => {
      await expect(Versioning.determineVersion(strategy, '')).rejects.toThrowErrorMatchingSnapshot();
    });

    describe('opt out strategy', () => {
      it("returns 'none'", async () => {
        await expect(Versioning.determineVersion('None', 'v1.0')).resolves.toMatchInlineSnapshot(`"none"`);
      });
    });

    describe('custom strategy', () => {
      test.each(['v0.1', '1', 'CamelCase', 'dashed-version'])(
        'returns the inputVersion for %s',
        async (inputVersion) => {
          await expect(Versioning.determineVersion('Custom', inputVersion)).resolves.toStrictEqual(inputVersion);
        },
      );
    });

    describe('semantic strategy', () => {
      it('refers to generateSemanticVersion', async () => {
        const generateSemanticVersion = jest.spyOn(Versioning, 'generateSemanticVersion').mockResolvedValue('1.3.37');

        await expect(Versioning.determineVersion('Semantic', '')).resolves.toStrictEqual('1.3.37');
        expect(generateSemanticVersion).toHaveBeenCalledTimes(1);
      });
    });

    describe('tag strategy', () => {
      it('refers to generateTagVersion', async () => {
        const generateTagVersion = jest.spyOn(Versioning, 'generateTagVersion').mockResolvedValue('0.1');

        await expect(Versioning.determineVersion('Tag', '')).resolves.toStrictEqual('0.1');
        expect(generateTagVersion).toHaveBeenCalledTimes(1);
      });
    });

    describe('not implemented strategy', () => {
      it('throws a not implemented exception', async () => {
        const strategy = 'Test';
        // @ts-ignore
        jest.spyOn(Versioning, 'strategies', 'get').mockReturnValue({ [strategy]: strategy });
        await expect(Versioning.determineVersion(strategy, '')).rejects.toThrowError(NotImplementedException);
      });
    });
  });

  describe('generateTagVersion', () => {
    it('removes the v', async () => {
      jest.spyOn(Versioning, 'getTag').mockResolvedValue('v1.3.37');
      await expect(Versioning.generateTagVersion()).resolves.toStrictEqual('1.3.37');
    });
  });

  describe('parseSemanticVersion', () => {
    it('returns the named parts', async () => {
      jest.spyOn(Versioning, 'getVersionDescription').mockResolvedValue('v0.1-2-g12345678');

      await expect(Versioning.parseSemanticVersion()).resolves.toMatchObject({
        tag: '0.1',
        commits: '2',
        hash: '12345678',
      });
    });

    it('throws when no match could be made', async () => {
      jest.spyOn(Versioning, 'getVersionDescription').mockResolvedValue('no-match-can-be-made');

      await expect(Versioning.parseSemanticVersion()).toMatchObject({});
    });
  });

  describe('getVersionDescription', () => {
    it('returns the commands output', async () => {
      const runOutput = 'someValue';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.getVersionDescription()).resolves.toStrictEqual(runOutput);
    });
  });

  describe('isShallow', () => {
    it('returns true when the repo is shallow', async () => {
      const runOutput = 'true\n';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.isShallow()).resolves.toStrictEqual(true);
    });

    it('returns false when the repo is not shallow', async () => {
      const runOutput = 'false\n';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.isShallow()).resolves.toStrictEqual(false);
    });
  });

  describe('fetch', () => {
    it('awaits the command', async () => {
      jest.spyOn(core, 'warning').mockImplementation(() => {});
      jest.spyOn(System, 'run').mockImplementation();
      await expect(Versioning.fetch()).resolves.not.toThrow();
    });

    it('falls back to the second strategy when the first fails', async () => {
      jest.spyOn(core, 'warning').mockImplementation(() => {});
      const gitFetch = jest.spyOn(System, 'run').mockImplementation();

      await expect(Versioning.fetch()).resolves.not.toThrow();
      expect(gitFetch).toHaveBeenCalledTimes(1);
    });
  });

  describe('generateSemanticVersion', () => {
    it('returns a proper version from description', async () => {
      jest.spyOn(System, 'run').mockImplementation();
      jest.spyOn(core, 'info').mockImplementation(() => {});
      jest.spyOn(Versioning, 'isDirty').mockResolvedValue(false);
      jest.spyOn(Versioning, 'hasAnyVersionTags').mockResolvedValue(true);
      jest.spyOn(Versioning, 'getTotalNumberOfCommits').mockResolvedValue(2);
      jest.spyOn(Versioning, 'parseSemanticVersion').mockResolvedValue({
        match: '0.1-2-g1b345678',
        tag: '0.1',
        commits: '2',
        hash: '1b345678',
      });

      await expect(Versioning.generateSemanticVersion()).resolves.toStrictEqual('0.1.2');
    });

    it('throws when dirty', async () => {
      jest.spyOn(System, 'run').mockImplementation();
      jest.spyOn(core, 'info').mockImplementation(() => {});
      jest.spyOn(Versioning, 'isDirty').mockResolvedValue(true);
      await expect(Versioning.generateSemanticVersion()).rejects.toThrowError();
    });

    it('falls back to commits only, when no tags are present', async () => {
      const commits = Math.round(Math.random() * 10);
      jest.spyOn(System, 'run').mockImplementation();
      jest.spyOn(core, 'info').mockImplementation(() => {});
      jest.spyOn(Versioning, 'isDirty').mockResolvedValue(false);
      jest.spyOn(Versioning, 'hasAnyVersionTags').mockResolvedValue(false);
      jest.spyOn(Versioning, 'getTotalNumberOfCommits').mockResolvedValue(commits);

      await expect(Versioning.generateSemanticVersion()).resolves.toStrictEqual(`0.0.${commits}`);
    });
  });

  describe('isDirty', () => {
    it('returns true when there are files listed', async () => {
      const runOutput = 'file.ext\nfile2.ext';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.isDirty()).resolves.toStrictEqual(true);
    });

    it('returns false when there is no output', async () => {
      const runOutput = '';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.isDirty()).resolves.toStrictEqual(false);
    });
  });

  describe('getTag', () => {
    it('returns the commands output', async () => {
      const runOutput = 'v1.0';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.getTag()).resolves.toStrictEqual(runOutput);
    });
  });

  describe('hasAnyVersionTags', () => {
    it('returns false when the command returns 0', async () => {
      const runOutput = '0';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.hasAnyVersionTags()).resolves.toStrictEqual(false);
    });

    it('returns true when the command returns >= 0', async () => {
      const runOutput = '9';
      jest.spyOn(System, 'run').mockResolvedValue(runOutput);
      await expect(Versioning.hasAnyVersionTags()).resolves.toStrictEqual(true);
    });
  });

  describe('getTotalNumberOfCommits', () => {
    it('returns a number from the command', async () => {
      jest.spyOn(System, 'run').mockResolvedValue('9');
      await expect(Versioning.getTotalNumberOfCommits()).resolves.toStrictEqual(9);
    });
  });
});
