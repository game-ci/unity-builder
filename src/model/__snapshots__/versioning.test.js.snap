// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Versioning determineVersion throws for invalid strategy 0 1`] = `"Versioning strategy should be one of None, Semantic, Tag, Custom."`;

exports[`Versioning determineVersion throws for invalid strategy somethingRandom 1`] = `"Versioning strategy should be one of None, Semantic, Tag, Custom."`;

exports[`Versioning determineVersion throws for invalid strategy undefined 1`] = `"Versioning strategy should be one of None, Semantic, Tag, Custom."`;
